[REEF-845] Make CommunicationGroupDriverImpl injectable

gyeongin opened the pull request:
This addressed the issue by
  * Make constructor of `CommunicationGroupDriverImpl`, `TreeTopology`, and `FlatTopology` injectable
  * Deprecate public constructors of `CommunicationGroupDriverImpl`, `TreeTopology`, and `FlatTopology`
  * Introduce `CommunicationGroupDriverFactory` and `TopologyFactory`
  * Fork injector to instantiate different communication groups and topologies

JIRA:
  [REEF-845](https://issues.apache.org/jira/browse/REEF-845)

tcNickolas commented:
Checkstyle comment: when you create a new class, please add class-level Javadoc to it. https://issues.apache.org/jira/browse/REEF-525 which will enforce this for reef-io project is not completed yet, but it will be.

gyeongin commented:
Thanks, @tcNickolas I've addressed your comments.

jsjason commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/CommunicationGroupDriverFactory.java:
```diff
@@ -0,0 +1,65 @@
+  private final Injector injector;
+
+  @Inject
+  private CommunicationGroupDriverFactory(final Injector injector) {
```
This factory needs `GroupCommSenderStage`, `DriverIdentifier`, `GroupCommRunningTaskHandler`, `GroupCommFailedTaskHandler`, and `GroupCommFailedEvalHandler` as additional configurations, which is not obvious. How about directly receiving those named parameters instead of an `Injector` object?

jsjason commented on lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/driver/TopologyFactory.java:
```diff
@@ -0,0 +1,55 @@
+  private final Injector injector;
+
+  @Inject
+  private TopologyFactory(final Injector injector) {
```
This factory needs `CommGroupNameClass`, `GroupCommSenderStage`, `DriverIdentifier`, `CommGroupNumTask`, and `TreeTopologyFanOut` as additional configurations, which is not obvious. How about directly receiving those named parameters instead of an `Injector` object?

jsjason commented:
@gyeongin, you've slightly changed how `RunningTask` and `FailedTask` events are broadcasted (e.g. there's no `commGroupRunningTaskHandler` now). Could you

jsjason commented:
@gyeongin Could you describe in detail what has changed? Some changes don't seem to be included in the pull request description.

gyeongin commented:
@jsjason Thanks for review. I've addressed your comments.

To describe the change related to event handlers, this PR removes redundant wrapping of comm group event handlers.
Current implementation of `CommunicationGroupDriverImpl` adds topology handlers to `commGroup*Handler`, and `GroupCommDriverImpl` adds `commGroup*Handler` to `groupComm*Handler`. For `GroupCommMessageHandler`, we distinguish `commGroupMessageHandler` of different comm groups and hand over `GroupCommunicationMessage` to specific `commGroupMessageHandler`. On the other hand, we do not distinguish specific comm group (running task, failed task, failed evaluator) handlers from others,  handing over all events to all comm group handlers, and thus this wrapping is redundant.

jsjason commented:
This looks good enough to merge. I'll test and merge.
